lua/ulx/modules/sh/cfc_curse.lua

```diff
@@ -52,13 +100,54 @@ function cmd.cursePlayers( callingPlayer, targetPlayers, effectName, durationMin
     end
 
     local durationSeconds = durationMinutes and durationMinutes * 60 or 0
+    local smallestAmount = amount
+    local largestAmount = 0
+
+    if amount then
+        -- Curse each person with multiple random effects. Doesn't allow uncursing.
+        for i = #targetPlayers, 1, -1 do
+            local plyAmount = amount
+            local ply = targetPlayers[i]
+            local availableEffects = CFCUlxCurse.FilterCompatibleEffects( ply, getOneTimeEffects() )
+
+            -- Only allow a given effect to be applied at most one time during the following loop.
+            -- This rule doesn't count for effects the player already has, allowing them to be refreshed one time. (new duration, new seed, etc.)
+
+            for _ = 1, amount do
+                local effectsLeft = #availableEffects
+
+                if effectsLeft == 0 then
+                    plyAmount = plyAmount - 1
+                else
+                    local effectInd = math.random( 1, effectsLeft )
+                    local effect = availableEffects[effectInd]
+                    local applyFailed = cmd.curse( ply, effect, durationSeconds, shouldUncurse )
+
+                    if applyFailed then
+                        plyAmount = plyAmount - 1
+                    else
+                        table.remove( availableEffects, effectInd )
+                        availableEffects = CFCUlxCurse.FilterCompatibleEffects( ply, availableEffects ) -- Filter out effects that might be incompatible with the new one.
+                    end
+                end
+            end
```

Comment:
This section is sketch.

```lua
availableEffects = CFCUlxCurse.FilterCompatibleEffects( ply, availableEffects ) -- Filter out effects that might be incompatible with the new one.
```
Is this necessary? `availableEffects` is already a filtered subset of the random effects you generated. Doesn't that mean that everything in `availableEffects` is compatible with each other _and_ the effects that already exist on the player?

If not, I think it should be.

If you cursed 40 players with 100 effects each, napkin maths says we're looking at like 200k total loop iterations across all the called functions, including table reshuffling from random-index removal.. bleh

It's probably fine, but there's gotta be a better way to do this.

Reply:
`availableEffects` only accounts for what curses the player can receive *right now*. When a new curse is given to a player, there's potentially new incompatibilities that would prevent them from receiving some of the other curses in `availableEffects`, so we need to update the list.

A great deal of curses have incompatibilities, so the available list shrinks quickly. The largest amount of effects a player can have at any one time is around 26 or 27, from testing. Once a player doesn't have any more curses left to receive, it just becomes a simple number decrement loop.